### PR TITLE
WIP: Adaptive continuation improvements

### PR DESCRIPTION
**Pull request by jameskermode**
Work in progress on improvements to add adaptive timestepping to the numerical continuation SinclairCrack class

### CODE REVIEW

File: matscipy/cauchy_born.py [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+            lattice2mask = force_diff_lattice[:,1]<0
+            if all(element == lattice1mask[0] for element in lattice1mask): #if they're all false, try the z component
+                lattice1mask = force_diff_lattice[:,2]>0
+                lattice2mask = force_diff_lattice[:,2]<0
+        
+
+        #set the masks for which atoms are in each of the sublattices
+        self.lattice1mask = lattice1mask
+        self.lattice2mask = lattice2mask     
+
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:101]: Style: only one blank line between methods in a class (two between functions in module). You could use a pep8 linter/formatter to fix automatically.

File: matscipy/cauchy_born.py [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+        
+
+    
+
+    def fit_taylor(self,de=1e-4):
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:106]: Style: always leave a space after comma-separated arguments

File: matscipy/cauchy_born.py [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+            using finite differences. Sets self.grad_f to the vector of first derivatives of the first shift component 
+            with each of the 6 strain components and self.hess_f to the hessian of the first shift component with the strain components.
+            NOTE this model is generally worse than the model provided later in initial_lasso_fit.
+        Parameters
+        ----------
+        de : float
+            tolerance for finite difference gradient calculations
+        '''
+        calc = self.calc
+        def f_gl(E_vec,calc,unitcell):
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:117]: pep8

File: matscipy/cauchy_born.py [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+            shift_diff[0] : float
+                first component of cauchy-born shift correction for applied strain state
+            '''
+            #green lagrange version of function
+            #turn E into matrix
+            E = np.zeros([3,3])
+            E[0,0] = E_vec[0]
+            E[1,1] = E_vec[1]
+            E[2,2] = E_vec[2]
+            E[2,1],E[1,2] = E_vec[3],E_vec[3]
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:139]: pep8: comma should be followed by space (later occurances not noted, but please fix them all!)

File: matscipy/cauchy_born.py [hunk 5]
```diff
@@ -0,0 +1,742 @@
...
+from scipy.linalg import polar, sqrtm
+
+from itertools import permutations
+
+from sklearn.linear_model import Lasso, LinearRegression
+
+from scipy.stats.qmc import LatinHypercube, scale
+
+class CubicCauchyBorn:
+    '''Corrector for the cauchy-born prediction of atomistic positions in multi-lattices subject to continuum strain
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:15]: pep8: please use `"""` for doc strings rather than `'''`

File: matscipy/cauchy_born.py [hunk 6]
```diff
@@ -0,0 +1,742 @@
...
+            F_3D = F_func(x,y,z,*args,**kwargs)
+            
+        else:
+            print('Coordinate system should be cart, cylind2D, cylind3D, spherical')
+            raise NotImplementedError
+
+        return F_3D
+
+    
+    def evaluate_F_or_E(self,A,atoms,F_func=None,E_func=None,cell=None,coordinates='cart3D',*args,**kwargs):
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:301]: There is not much common code between the F and E cases. Maybe better to split into two functions, once which evaluates F and one that evaluates E?

File: matscipy/cauchy_born.py [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+
+from ase.optimize import LBFGS
+
+from scipy.linalg import polar, sqrtm
+
+from itertools import permutations
+
+from sklearn.linear_model import Lasso, LinearRegression
+
+from scipy.stats.qmc import LatinHypercube, scale
```
Comment (jameskermode) [inline matscipy/cauchy_born.py:12]: Currently matscipy does not depend on sklearn. Do we really need this? If so it should be an optional dependency: try to import it, and if it fails raise an exception later when someone tries to use the functionality.

File: matscipy/fracture_mechanics/clusters.py [hunk 8]
```diff
@@ -130,3 +170,57 @@ def bcc(*args, **kwargs):
...
+
+    # Find surface energy
+    import numpy as np
+    Ebulk = bulk.get_potential_energy() ; Eslab = slab.get_potential_energy()
+    area = np.linalg.norm(np.cross(slab.get_cell()[1,:],slab.get_cell()[2,:]))
+    gamma_ase = (Eslab - Ebulk)/(2*area)
+
+    # Convert to required units
+    if unit == 'ASE':
+        return [gamma_ase,'ase_units']
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/clusters.py:217]: Results should always be a tuple not a list.

Also, returning unit as a string is a bit odd. I suggest just return the number, since the user knows which unit system they asked for.

File: matscipy/fracture_mechanics/clusters.py [hunk 9]
```diff
@@ -130,3 +170,57 @@ def bcc(*args, **kwargs):
...
+    else:
+        from ase import units
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/clusters.py:219]: move import to top of file

File: matscipy/fracture_mechanics/crack.py [hunk 10]
```diff
@@ -54,6 +62,208 @@
 MPa_sqrt_m = 1e6*units.Pascal*np.sqrt(units.m)
 
 ###
+# Temporary function. Eventually meant to be pushed to ase.optimize.ode
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:65]: Since this will not happen fast maybe we should move it to a new `matscipy.optimize` module?

File: matscipy/fracture_mechanics/crack.py [hunk 11]
```diff
@@ -359,7 +569,8 @@ class CubicCrystalCrack:
     """
 
     def __init__(self, crack_surface, crack_front, C11=None, C12=None,
-                 C44=None, stress_state=PLANE_STRAIN, C=None, Crot=None):
+                 C44=None, stress_state=PLANE_STRAIN, C=None, Crot=None,
+                 cb=None):
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:573]: `cb` is not an obvious name: could we spell out in full, e.g. `cauchy_born`?

File: matscipy/fracture_mechanics/crack.py [hunk 12]
```diff
@@ -715,7 +932,7 @@ class SinclairCrack:
     def __init__(self, crk, cryst, calc, k, alpha=0.0, vacuum=6.0,
                  variable_alpha=True, variable_k=False,
                  alpha_scale=None, k_scale=None,
-                 extended_far_field=False):
+                 extended_far_field=False,rI=0.0,rIII=0.0,cutoff=0.0):
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:935]: pep8: space after comma

File: matscipy/fracture_mechanics/crack.py [hunk 13]
```diff
@@ -836,6 +1062,98 @@ def u_cle(self, alpha=None):
         u = np.c_[ux, uy, np.zeros_like(ux)] # convert to 3D field
         return u
 
+
+    #-----------------delete this part from main branch-------------------
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:1066]: action required?

File: matscipy/fracture_mechanics/crack.py [hunk 14]
```diff
@@ -907,13 +1276,13 @@ def set_atoms(self, atoms):
         self.update_atoms()  # now we have same u_cle in atoms and self.atoms
         min_len = min(N1_in, self.N1)
         # FIXME this assumes stable sort order for atoms and self.atoms
-        u = atoms.positions[:min_len] - self.atoms.positions[:min_len]
+        u = atoms.positions - self.atoms.positions
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:1279]: why is this change needed?

File: matscipy/fracture_mechanics/crack.py [hunk 15]
```diff
@@ -989,10 +1362,18 @@ def get_forces(self, x1=None, xdot1=None, ds=None, forces=None, mask=None):
...
+            self.f_alpha_correction = 0.0 #self.get_f_alpha_correction()
+            F.append((f_alpha-self.f_alpha_correction)) #ninsert some kind of botched preconditioning.
+            print('f_alpha',f_alpha)
+            print('corrected f_alpha',(f_alpha-self.f_alpha_correction))
         if self.variable_k:
             f_k = self.get_k_force(x1, xdot1, ds)
             F.append(f_k)
+            print('f_k', f_k)
+
+        #self.norm_F_vals.append(np.linalg.norm(F,np.inf))
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:1374]: cleanup unneeded code

File: matscipy/fracture_mechanics/crack.py [hunk 16]
```diff
@@ -1087,6 +1468,11 @@ def log(x, f=None):
 
         def residuals(x, *args):
             self.set_dofs(x)
+            #self.write_atoms_to_file() #comment out at some point
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:1471]: cleanup or add an optional argument to enable this output

File: matscipy/fracture_mechanics/crack.py [hunk 17]
```diff
@@ -1269,21 +1694,47 @@ def arc_length_continuation(self, x0, x1, N=10, ds=0.01, ftol=1e-2,
...
-            # monitor sign of \dot{alpha} and flip if necessary
             if self.variable_alpha:
-                _, alphadot2, _ = self.unpack(xdot2)
+                # monitor sign of \dot{alpha} and flip if necessary
+                #udot1, alphadot1, kdot1 = self.unpack(xdot1)
+                udot2, alphadot2, kdot2 = self.unpack(xdot2)
                 if direction * np.sign(alphadot2) < 0:
-                    xdot2 = -xdot2
-
+                    xdot2 = -xdot2
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:1726]: cleanup

File: matscipy/fracture_mechanics/crack.py [hunk 18]
```diff
@@ -1387,20 +1859,217 @@ def frame(idx):
...
             tip.set_data([self.cryst.cell[0, 0] / 2.0 + x[idx, -2]],
                          [self.cryst.cell[1, 1] / 2.0])
             return blob, plot_elements, lc, tip
 
         if frames is None:
             frames = range(0, len(x), 100)
+
         return FuncAnimation(fig, frame, frames)
 
+    @counted
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:1878]: what does this  `@counted` decorator do?

File: matscipy/fracture_mechanics/crack.py [hunk 19]
```diff
@@ -1387,20 +1859,217 @@ def frame(idx):
...
+        #print(J)
+        v = self.u
+        dv = np.linalg.norm(v[I, :] - v[J, :], axis=1)
+        r = self.r[self.regionI][I]
+        mask = r<(self.rI-cutoff)
+        if seperate_surface:
+            mid_cell_x = self.cryst.cell.diagonal()[0] / 2.0
+            mid_cell_y = self.cryst.cell.diagonal()[1] / 2.0
+            I_positions = self.cryst.get_positions()[I]
+            x_criteria = ((I_positions[:,0]-mid_cell_x)<0.5)
```
Comment (jameskermode) [inline matscipy/fracture_mechanics/crack.py:2000]: pep8: spaces around binary operator (`-` in this case)

File: matscipy/surface_reconstruction.py [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+from matscipy.cauchy_born import CubicCauchyBorn
+from ase.optimize.precon import PreconLBFGS
+import ase.io
+
+
+
+class SurfaceReconstruction:
+    def __init__(self,el,a0,calc,directions,surf_dir,lattice='diamond'):
+
```
Comment (jameskermode) [inline matscipy/surface_reconstruction.py:16]: doc string still to come

File: matscipy/fracture_mechanics/crack.py (hunk 18 above)
Comment (Fraser-Birks) [reply matscipy/fracture_mechanics/crack.py:1878]: @counted just keeps track of how many times the write_atoms_to_file function is called, this allows the function to save .xyz files with consecutive names each time it is called. Admittedly, I now realise the same effect could have been achieved by having an instance attribute for sinclair_crack which keeps track of how often the function is called

File: matscipy/fracture_mechanics/crack.py (hunk 14 above)
Comment (Fraser-Birks) [reply matscipy/fracture_mechanics/crack.py:1279]: This was a necessary change for the case of having an extended region I which encompassed the surfaces. The original code only worked when region I was made up of the first min_len atoms, the new code works for a region I of any shape.

**Comment by jameskermode** [comment]
Neighbour list test failure is fixed on master - if you merge that into this branch then tests should pass

File: matscipy/fracture_mechanics/crack.py (hunk 14 above)
Comment (jameskermode) [reply matscipy/fracture_mechanics/crack.py:992]: ok

File: matscipy/fracture_mechanics/crack.py (hunk 18 above)
Comment (jameskermode) [reply matscipy/fracture_mechanics/crack.py:1581]: ok - not too fussed by an explicit instance attribute might be clearer
